### DISCUSSION TITLE
auth.lower + shorttitle(3,3) + year will cause citation key too long.

### DISCUSSION THREAD

**Discussion by earthmoon** [retorquere/zotero-better-bibtex#2362] [Q&A]
Hello,
If using the default setting of `auth.lower + shorttitle(3,3) + year` in the Better Bibtex, there are two problems:

1. The citation key will be very long, as it would use all the Chinese characters in title;
2. No matter whether enabling the function of `Force citation key to plain text`, it always converts Chinese character in the title to plain text.

If the reference contains Chinese characters, the citation key will affect any thing in Latex. Because the Latex file of course can deal with Chinese characters.

**Comment by retorquere** [comment]
Use `auth.lower + title.jieba.select(n=3) + year`

**Comment by earthmoon** [comment]
Isn’t jieba Pinyin? Is it about Japanese? If it can use Pinyin, it is better. Thanks.
________________________________

**Reply by retorquere** [reply]
Use `auth.lower + title.jieba.fold(chinese).select(n=3) + year`

**Reply by earthmoon** [reply]
> auth.lower + title.jieba.fold(chinese).select(n=3) + year

Hi retorquere,
I have tested two methods of `auth.lower + title.jieba.select(n=3) + year` and `auth.lower + title.jieba.fold(chinese).select(n=3) + year`. The first will use all the Chinese characters of titles and the second will convert all the Chinese characters to Pinyin.  They are all too long and `n=3` doesn't work.

**Reply by JHC-NEWGUY** [reply]
1. Thank you for your reply and guidance.
2. I guess it's a problem of my own ability. Because I'm not a programmer myself and only know a little knowledge. Now I use this code `[auth][year][shortitle]`, and the output result is what I said above.
3. After your prompt, I found that it is okay to use this code. Thank you very much.

Code: auth(n=0,m=1,creator="*",initials=false).fold + year + shorttitle.jieba.fold(chinese).select(0,3)

**Comment by JHC-NEWGUY** [comment]
我就直接使用中文了，我估计老外应该不会遇到这个问题。
结论：个人认为无解
如果是写中文的期刊论文，一般不会使用latex。（毕业论文除外）
如果是写英文的期刊论文，一般也不会引用中文的文献，所以不存在这个问题。
****
1. 我想，那个东西之所以长，主要的原因是在中文方面。譬如“基于特征筛选和集成学习的轴承故障诊断”，被弄成了“chenshengfan2023JiYuTeZhengShaiXuanHeJiChengXueXiDeZhouChengGuZhangZhenDuan”，英文的文献，应该是没有这个问题的。
2. 我去查了下，https://retorque.re/zotero-better-bibtex/citing/，也可以看下，select的用法是，从传入的值中选择单词。格式为select(start,number)（从 1 开始），所以select(1,4)或select(n=4)将选择前四个单词。如果未给出，则选择从到结尾的所有单词。注意是单词。那一堆中文的拼音在外文里面被默认问了一个单词，所以，没有什么用处的。
3. 整个看了下，没有截断这个操作的，所以放弃了，就这样吧。或者等大牛了

**Reply by earthmoon** [reply]
你好，
一门语言没有截断操作吗？
谢谢！

**Reply by retorquere** [reply]
If my involvement is required it'll have to be in English.

**Reply by JHC-NEWGUY** [reply]
1. I'm so sorry. I really didn't expect you to reply.
2. Thank you very much for this program you made. It's very useful. Thank you so much.
3. I felt this requirement wasn't necessary, so I used Chinese.

The situation is like this:
If it's an English title, each word has a space. In this case, if using Better BibTeX, there is no problem.
For example: Fault Diagnosis of Bearing in Wind Turbine Gearbox Under Actual Operating Conditions Driven by Limited Data With Noise Labels
The above title has more than 10 words. The first word is "Fault"; the second word is "Diagnosis"; the third word is "Bearing".
If using Better BibTeX, it can be selected through shorttitle(1)=Fault or shorttitle(3)=FaultDiagnosisBearing.
****
However, if it's translated into Chinese: 带噪声标签的有限数据驱动下实际运行条件下风力涡轮机齿轮箱轴承的故障诊断. When using Better BibTeX, it has been transformed into pinyin letters: daizaoshengbiaoqiandeyouxianshujuqudongxiashijiyunxingtiaojianxiafengliwolunjichilunxiangzhouchengdeguzhangzhenduan
In the above pinyin, there are no spaces. If viewed in the form of English, it's just one long word, which is the above long pinyin.
The questioner and I hope to have a function, that is, a function like left(6) in excel can be used.
For instance: left(6) = daizao; left(9) = daizaoshe; such a function. That is, I can choose the length of the word by myself.

**Reply by retorquere** [reply]
Use something like `title.jieba.select`.

**Reply by retorquere** [reply]
Or `.substring`, but `.jieba` sounds closer to what you describe you want.

**Reply by JHC-NEWGUY** [reply]
1. Thank you for your reply and guidance.
2. I guess it's a problem of my own ability. Because I'm not a programmer myself and only know a little knowledge. Now I use this code `[auth][year][shortitle]`, and the output result is what I said above.
3. After your prompt, I found that it is okay to use this code. Thank you very much.

Code: auth(n=0,m=1,creator="*",initials=false).fold + year + shorttitle.jieba.fold(chinese).select(0,3)

**Reply by JHC-NEWGUY** [reply]
Reminder:
1. I am using Zotero version 6.0.36.
2. `auth.lower + year + title.jieba.fold(chinese).select(n=2)` is also okay.
3. `[auth.lower][year][title.jieba.fold(chinese).select(n=2)]` is also okay.

**Reply by retorquere** [reply]
No problem, it's always OK to ask.

The bracketed syntax is supported but doesn't support everything of the new format.

**Comment by kerim** [comment]
So, I've been using the formula

auth.lower + shorttitle(3, 3) + year

And ran into this issue with a Chinese entry. The title is: 

「離開台北就成異鄉人，對城市外的社會充滿興趣」：台灣影像人類學誌先驅胡台麗病逝，享壽72歲 

and the Citation Key is 

thenewslensLiKaiTaiBeiJiuChengYiXiangRenDuiChengShiWaiDeSheHuiChongManXingQuTaiWanYingXiangRenLeiXueZhiXianQuHuTaiLiBingShiXiangShou72SuiOnceYou2022

It seems that I can use the above mentioned formulas to make it shorter, but I'm worried about replacing my formula, because I don't want to mess with my other citation keys which were made with the previous formula.

How do I manage this?

**Reply by earthmoon** [reply]
use this `auth.lower + title.jieba.select(n=3) + year`

**Reply by kerim** [reply]
Thanks, but after replacing the formula, nothing changed. I clicked to refresh the Cite Key, but it is still the same.

**Reply by earthmoon** [reply]
> Thanks, but after replacing the formula, nothing changed. I clicked to refresh the Cite Key, but it is still the same.

try this:
`language(chinese) + authors(1, name='%(f)s%(g)s') + year | auth.lower + year`

**Comment by kerim** [comment]
Thanks. It is very finicky. 

This finally worked for me:

```
language(en) + auth.lower + shorttitle(3, 3) + year
language(zh) + authors(1, name = "%(f)s%(g)s") + year | auth.lower + year
```

various other arrangements didn't work:
- Listing language(zh) first
- removing language(en) from the default
- writing "Chinese" instead of "zh"
etc. all broke it, but somehow this particular format works.

**Reply by retorquere** [reply]
> writing "Chinese" instead of "zh" etc. all broke it,

That was a bug. Please report breakage. A new release build is running.

**Reply by retorquere** [reply]
7.0.61 is out
